### PR TITLE
UIA: use full buffer comparison in rects and endpoint setter

### LINKED ISSUE #6402: Narrator scan mode crashes terminal
# Environment

Windows build number: Microsoft Windows [Version 10.0.18363.836]
Windows Terminal version (if applicable): 1.0.1401.0

Any other software?: Narrator, the version which comes with this windows build

# Steps to reproduce

* Start Narrator using Win+Ctrl+Enter
* Open the Windows Terminal app
  * Make sure the Narrator cursor (blue box) is on the terminal cursor, and/or that Narrator says "editing" as part of the information when you navigate to the terminal window (if Narrator is not focused on the correct place in the terminal window, alt-tabbing away and back should fix it).
* Activate scan mode with Caps+Space (Narrator will say "scan")
* Press the down arrow key (Narrator will say "title bar")
  * If Narrator does not announce "title bar", it may be further down in the UI tree. Keep pressing the down arrow until you reach it
* Press the up arrow key, and the Windows Terminal app crashes

# Expected behavior

I would expect the terminal app not to crash, and for Narrator focus to return to the edit box.

# Actual behavior

The terminal app crashes.

# Feedback hub link

https://aka.ms/AA8mlsm

### PR DESCRIPTION
**Pull request by carlos-zamora**
In UiaTextRange, `_getBufferSize` returns an optimized version of the
size of the buffer to be the origin and the last character in the
buffer. This is to improve performance on search or checking if you are
currently on the last word/line.

When setting the endpoint and drawing the bounding rectangles, we should
be retrieving the true buffer size. This is because it is still possible
to create UiaTextRanges that are outside of this optimized size. The
main source of this is `ExpandToEnclosingUnit()` when the unit is
`Document`. The end _should_ be the last visible character, but it isn't
because that would break our tests.

This is an incomplete solution. #6986 is a follow up to completely test
and implement the solution.

The crash in #6402 was caused by getting the document range (a range of
the full text buffer),  then moving the end by one character. When we
get the document range, we get the optimized size of the buffer (the
position of the last character). Moving by one character is valid
because the buffer still has more to explore. We then crash from
checking if the new position is valid on the **optimized size**, not the
**real size**.

REFERENCES

#6986 - follow up to properly handle/test this "end of buffer" problem

Closes #6402

### CODE REVIEW

**Comment by DHowett** [comment]
nit: I'd like the title to explain _when_, not just say "sometimes"

File: src/types/UiaTextRangeBase.cpp [hunk 1]
```diff
@@ -393,7 +395,9 @@ IFACEMETHODIMP UiaTextRangeBase::GetBoundingRectangles(_Outptr_result_maybenull_
         // set of coords.
         std::vector<double> coords;
 
-        const auto bufferSize = _getBufferSize();
+        // GH#6402: Get the actual buffer size here, instead of the one
+        //          constructed by the virtual bottom.
```
Comment (DHowett) [inline src/types/UiaTextRangeBase.cpp:399]:
```suggestion
        //          constrained by the virtual bottom.
```

File: src/types/UiaTextRangeBase.cpp [hunk 2]
```diff
@@ -145,7 +145,9 @@ const COORD UiaTextRangeBase::GetEndpoint(TextPatternRangeEndpoint endpoint) con
 // - true if range is degenerate, false otherwise.
 bool UiaTextRangeBase::SetEndpoint(TextPatternRangeEndpoint endpoint, const COORD val) noexcept
 {
-    const auto bufferSize = _getBufferSize();
+    // GH#6402: Get the actual buffer size here, instead of the one
+    //          constructed by the virtual bottom.
```
Comment (DHowett) [inline src/types/UiaTextRangeBase.cpp:149]:
```suggestion
    //          constrained by the virtual bottom.
```

**Comment by codeofdusk** [comment]
Will commenting here prevent an automatic close?

**Comment by codeofdusk** [comment]
Might a similar approach be needed to fix #5481?

**Comment by codeofdusk** [comment]
@carlos-zamora Any updates on this PR?

**Comment by carlos-zamora** [comment]
> @carlos-zamora Any updates on this PR?

Hey, sorry. I think I'm going to have to reevaluate some of my open accessibility PRs today. Both of them have to do with defining the end of the document to be the last character. Which means we'll have to update the tests too. I'm leaving this open for now, but expect a better plan later today or tomorrow :) Sorry for the delay!

**Comment by DHowett** [comment]
What cases does this open up? When we have an endpoint beyond the visible screen, _what happens_? What if you move it more? What if you read the range aloud, get its text?

**Comment by carlos-zamora** [comment]
> 1. Please change the name of this PR.
✔

> 2. Your comments explain the what, not the why
> 3. I still don't understand the why
✔

> 4. So, it's possible to navigate character-by-character off the bottom of the viewport? **what happens after that**?
Navigation still uses the optimized buffer size. So it gets clamped to the populated text area, then moves.

> 5. What happens when you do this at the actual bottom of the buffer at 9001 lines?
This is handled the same as above.

> What cases does this open up? When we have an endpoint beyond the visible screen, what happens? What if you move it more? What if you read the range aloud, get its text?

Scrolling is of no concern because TermCore clamps the scroll offset to be valid.

The text is still extracted properly as you are only extracting spaces when in the non-populated region of the buffer.

The main concern is that this lacks polish. We should provide more tests (as you requested earlier), but I've moved that work into a follow-up item to ensure we are treating the end of the buffer correctly and wholistically.

**Comment by codeofdusk** [comment]
> The end should be the last visible character, but it isn't
because that would break our tests.

Is there any chance that the end *could* be made the last visible character here? Or is that the aim of #6986?

**Comment by DHowett** [comment]
That's the aim of #6986. :smile:

**Pull request merged**